rfcs/0007-pass-spec.md

```diff
@@ -0,0 +1,473 @@
+
+Furthermore, each pass is associated with a context type (`EventCtx`, `LifecycleCtx`, `LayoutCtx`, etc) which provides methods to access the environment.
+Some of these methods are shared between all passes, some by all but one pass, and some are specific to a single pass.
+Why a given method is available in one pass and not another is currently undocumented; it's obvious in some cases (e.g. it makes sense that EventCtx would have a `set_handled` method and PaintCtx wouldn't), but even then we should have a general model what capabilities are available in which passes.
+
+
+## User-facing explanation
+
+Masonry has a set of **passes**, which are computations run over a subset of the widget tree during a frame.
+
+
+### Event passes
+
+When a user interacts with the application in some way, like a mouse click, Masonry runs an **event pass** over the tree.
+There are three types of event passes:
+
+- **on_pointer_event:** covers positional events from the mouse and other pointing devices (pen, stylus, touchpad, etc).
+- **on_text_event:** text input events like keyboard presses, IME, clipboard paste, etc.
+- **on_access_event:** events from the OS's accessibility API.
+
+When an event occurs, the application selects the widget targeted by the event.
+For pointer events, this is either the widget under the pointer or the widget with pointer capture.
+For text and accessibility events, this is the widget with focus.
+
+The widget's event handling method (`on_pointer_event`, `on_text_event`, or `on_access_event`) is called.
+Then, the same method is called for each of the widget's parents, up to the root.
+This behavior is known in browsers as event bubbling.
+
+
+### Rewrite passes
+
+After an event pass, some flags may have been changed, and some values may have been invalidated and need to be recomputed.
+To address these invalidations, Masonry runs a set of **rewrite passes** over the tree:
+
+- **MUTATE** pass.
+- **UPDATE_TREE** pass.
+- **UPDATE_FOCUS** pass.
+- **UPDATE_DISABLED** pass.
+- **UPDATE_ANIM** pass.
+- **layout** pass.
+- **UPDATE_SCROLLS** pass.
+- **compose** pass.
+- **UPDATE_POINTER** pass.
+
+(The lowercase passes have methods with matching names in the Widget trait.)
+
+By default, each of these passes returns immediately, unless pass-dependent invalidation flags are set or work is requested.
+Each pass can generally request work for later passes; for instance, the MUTATE pass can invalidate the layout of a widget, in which case the `layout` pass will run on that widget and its children and parents.
+
+Passes may also request work for *previous* passes, in which case all rewrite passes are run again in sequence.
+For instance, the UPDATE_POINTER may change a widget's size, requiring another layout pass.
+
+To avoid infinite loops in those cases, the number of reruns has a static limit.
+If passes are still requested past that limit, they're delayed to a later frame.
```

Comment:
I guess that's relevant for frame-pacing, would it make sense to instead use a max duration limit (possibly relevant for layout too)?
(infinite-loop detection would be even better I guess, but that's out of scope for now I guess)

Reply:
The reason for this design is that the only infinite loops would be pathological cases, and we expect any reasonable scene to only require 2 or maybe 3 cycles. That is, it's a bug in your app if you hit this fallback, so it's not a problem to have it present.